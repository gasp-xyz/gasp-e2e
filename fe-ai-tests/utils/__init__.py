# This file makes the utils directory a Python package
# This allows imports like: from utils.common_test_utils import x
