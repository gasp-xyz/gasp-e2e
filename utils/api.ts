import { ApiPromise } from "@polkadot/api";
import { testLog } from "./Logger";
import { getEnvironmentRequiredVars, getMangataApiUrlPort } from "./utils";
import { Mangata, MangataInstance } from "@mangata-finance/sdk";
import { getPort } from "get-port-please";
import XcmNetworks from "./Framework/XcmNetworks";
import { BuildBlockMode } from "@acala-network/chopsticks";
import { ApiContext } from "./Framework/XcmHelper";

export let api: ApiPromise | null = null;
export let mangata: MangataInstance | null = null;
export let mangataChopstick: ApiContext | null = null;
export let chopstickUri: String;
export const getApi = () => {
  if (!api) {
    throw new Error("Api not initialized");
  }
  return api;
};

export const initApi = async (uri = "") => {
  if (!uri) {
    const { chainUri: envUri } = getEnvironmentRequiredVars();
    uri = envUri;
  }
  if (process.env.CHOPSTICK_ENABLED) {
    const mgaPort = getMangataApiUrlPort();
    const chopstickPort = await getPort();
    mangataChopstick = await XcmNetworks.mangata({
      localPort: chopstickPort,
      buildBlockMode: BuildBlockMode.Instant,
    });
    uri = uri.replace(mgaPort.toString(), chopstickPort.toString());
    chopstickUri = uri;
  }
  testLog.getLog().info(`TEST_INFO: Running test in ${uri}`);
  mangata = Mangata.instance([uri]);
  api = await mangata.api();
  return api;
};

export async function getMangataInstance(uri = ""): Promise<MangataInstance> {
  if (!api) {
    await initApi(uri);
  }
  return mangata!;
}
export const disconnect = async () => {
  //TODO:sdkV2
  //  if (mangata) {
  //    await mangata.disconnect();
  //  } else {
  //    await Mangata.instance([uri]).disconnect();
  //  }
};
