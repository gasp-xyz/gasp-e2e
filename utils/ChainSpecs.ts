import { BN_TEN, BN_ZERO } from "@mangata-finance/sdk";
import { BN, BN_HUNDRED, BN_THOUSAND } from "@polkadot/util";
import { ApiPromise } from "@polkadot/api";

export const TRANSFER_INSTRUCTIONS = 4;
export const WEIGHT_IN_SECONDS = new BN(1_000_000_000_000);

export interface ChainSpec {
  unitCostWeight: number;
  parachain: number;
  assets: Map<AssetSpec, AssetMeta>;
  foreign: Map<AssetSpec, AssetMeta>;
}

export interface AssetSpec {
  symbol: string;
  decimals: number;
  location: object;
  unit: BN;
}

interface AssetMeta {
  fps: BN;
  ed: BN;
  location: object;
}

export enum ChainId {
  Mg,
  Tur,
  Bifrost,
  Statemine,
}

export class AssetId {
  static Tur: AssetSpec = {
    symbol: "TUR",
    decimals: 10,
    location: { parents: 1, interior: { X1: { Parachain: 2114 } } },
    unit: BN_TEN.pow(new BN(10)),
  };
  static TurV3: AssetSpec = {
    symbol: "TUR",
    decimals: 10,
    location: { parents: 1, interior: { X1: { Parachain: 2114 } } },
    unit: BN_TEN.pow(new BN(10)),
  };

  static Mgx: AssetSpec = {
    symbol: "MGX",
    decimals: 18,
    location: {
      parents: 1,
      interior: { X2: [{ Parachain: 2110 }, { GeneralKey: "0x00000000" }] },
    },
    unit: BN_TEN.pow(new BN(18)),
  };
  static async getLocationFromChain(symbol: string, api: ApiPromise) {
    const entries = await api.query.assetRegistry.metadata.entries();
    const asset = entries.filter((entry) => {
      const elem = entry[1];
      const name = api.createType(
        "Vec<u8>",
        JSON.parse(JSON.stringify(elem)).symbol,
      );
      return name.toHuman() === symbol;
    })[0];
    return JSON.parse(JSON.stringify(asset[1].toHuman())).location;
  }
  static Bnc: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 1,
      interior: { X2: [{ Parachain: 2001 }, { GeneralKey: "0x0001" }] },
    },
    unit: BN_TEN.pow(new BN(12)),
  };
  static ImbueBncV3: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 0,
      interior: {
        X1: [
          {
            GeneralKey: {
              length: 2,
              data: "0x0001000000000000000000000000000000000000000000000000000000000000",
            },
          },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(12)),
  };
  static BncV3: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: 2001 },
          {
            GeneralKey: {
              length: 2,
              data: "0x0001000000000000000000000000000000000000000000000000000000000000",
            },
          },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(12)),
  };

  static USDt: AssetSpec = {
    symbol: "USDt",
    decimals: 6,
    location: {
      parents: 1,
      interior: {
        X3: [
          { Parachain: 1000 },
          { PalletInstance: 50 },
          { GeneralIndex: 1984 },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(6)),
  };
}

export const ChainSpecs = new Map<ChainId, ChainSpec>([
  [
    ChainId.Tur,
    {
      unitCostWeight: 1_000_000_000,
      parachain: 2114,
      assets: new Map([
        [
          AssetId.Tur,
          {
            fps: new BN(416_000_000_000),
            ed: AssetId.Tur.unit.mul(BN_TEN),
            location: AssetId.Tur.location,
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
  [
    ChainId.Mg,
    {
      unitCostWeight: 150_000_000,
      parachain: 2110,
      assets: new Map(),
      foreign: new Map([
        [
          AssetId.Tur,
          {
            fps: new BN(537_600_000_000),
            ed: BN_ZERO,
            location: AssetId.Tur.location,
          },
        ],
        [
          AssetId.TurV3,
          {
            fps: new BN(537_600_000_000),
            ed: BN_ZERO,
            location: AssetId.TurV3.location,
          },
        ],
        [
          AssetId.Bnc,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.Bnc.location,
          },
        ],
        [
          AssetId.ImbueBncV3,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.BncV3.location,
          },
        ],
        [
          AssetId.BncV3,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.BncV3.location,
          },
        ],
        [
          AssetId.USDt,
          {
            fps: new BN(13440000),
            ed: BN_ZERO,
            location: AssetId.USDt.location,
          },
        ],
      ]),
    },
  ],
  [
    ChainId.Bifrost,
    {
      unitCostWeight: 200_000_000,
      parachain: 2001,
      assets: new Map([
        [
          AssetId.Bnc,
          {
            fps: new BN(9360000000000),
            ed: AssetId.Bnc.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: { X1: { GeneralKey: "0x0001" } },
            },
          },
        ],
        [
          AssetId.BncV3,
          {
            fps: new BN(9360000000000),
            ed: AssetId.BncV3.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0001000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
        ],
        [
          AssetId.ImbueBncV3,
          {
            fps: new BN(9360000000000),
            ed: AssetId.ImbueBncV3.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0001000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
  [
    ChainId.Statemine,
    {
      unitCostWeight: 1_000_000_000,
      parachain: 1000,
      assets: new Map([
        [
          AssetId.USDt,
          {
            fps: new BN(1),
            ed: AssetId.USDt.unit.div(BN_THOUSAND),
            location: {
              parents: 0,
              interior: {
                X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
              },
            },
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
]);
