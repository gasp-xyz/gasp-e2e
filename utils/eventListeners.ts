/* eslint-disable no-loop-func */
import { MangataGenericEvent } from "@mangata-finance/sdk";
import { ApiPromise } from "@polkadot/api";
import { BN } from "@polkadot/util";
import _, { reject } from "lodash-es";
import { getApi, getMangataInstance } from "./api";
import { logEvent, testLog } from "./Logger";
import { api } from "./setup";
import { getEventErrorFromSudo } from "./txHandler";
import { User } from "./User";
import { getEnvironmentRequiredVars } from "./utils";
import { Codec } from "@polkadot/types/types";

// lets create a enum for different status.
export enum ExtrinsicResult {
  ExtrinsicSuccess,
  ExtrinsicFailed,
  ExtrinsicUndefined,
}

///Class that stores the event result.
export class EventResult {
  /**
   *
   */
  constructor(
    state: ExtrinsicResult = ExtrinsicResult.ExtrinsicUndefined,
    data: any,
  ) {
    this.state = state;
    this.data = data;
  }

  state: ExtrinsicResult;
  data: String;
}

// for testing
export const getEventResult = (
  section: any,
  method: any,
  module_index: any,
) => {
  const api = getApi();

  return new Promise<EventResult>(async (resolve) => {
    const unsubscribe = (await api.query.system.events((events: any) => {
      events.forEach((record: any) => {
        const { event } = record;
        if (event.section === section && event.method === method) {
          unsubscribe();
          resolve(
            new EventResult(
              ExtrinsicResult.ExtrinsicSuccess,
              JSON.parse(event.data.toString()),
            ),
          );
        } else if (
          event.section === "system" &&
          event.method === "ExtrinsicFailed" &&
          (JSON.parse(event.data.toString())[0].Module.index = module_index)
        ) {
          unsubscribe();
          resolve(
            new EventResult(
              ExtrinsicResult.ExtrinsicFailed,
              JSON.parse(event.data.toString())[0].Module.error,
            ),
          );
        }
      });
    })) as any;
  });
};

export const waitNewBlock = () => {
  const api = getApi();
  let count = 0;
  return new Promise(async (resolve) => {
    const unsubscribe = await api.rpc.chain.subscribeNewHeads((header: any) => {
      if (++count === 2) {
        testLog.getLog().info(`Chain is at block: #${header.number}`);
        unsubscribe();
        resolve(true);
      }
    });
  });
};

export function filterEventData(
  result: MangataGenericEvent[],
  method: string,
): any[] {
  return result
    .filter((event) => `${event.section}.${event.method}` === method)
    .map((event) => event.event.toHuman().data);
}

export function findEventData(result: MangataGenericEvent[], method: string) {
  return filterEventData(result, method)[0];
}

export async function waitSudoOperationSuccess(
  checkingEvent: MangataGenericEvent[],
) {
  const filterBootstrapEvent = checkingEvent.filter(
    (extrinsicResult) => extrinsicResult.method === "Sudid",
  );

  const userBootstrapCall = filterBootstrapEvent[0].event.data[0].toString();

  expect(userBootstrapCall).toContain("Ok");
}

export async function waitSudoOperationFail(
  checkingEvent: MangataGenericEvent[],
  expectedErrors: string[],
) {
  const filterBootstrapEvent = checkingEvent.filter(
    (extrinsicResult) => extrinsicResult.method === "Sudid",
  );

  const BootstrapError = await getEventErrorFromSudo(filterBootstrapEvent);

  expect(expectedErrors).toContain(BootstrapError.method);
}

export const waitForEvents = async (
  api: ApiPromise,
  method: string,
  blocks: number = 10,
  withData: string = "",
): Promise<CodecOrArray> => {
  return new Promise(async (resolve, reject) => {
    let counter = 0;
    const unsub = await api.rpc.chain.subscribeFinalizedHeads(async (head) => {
      const events = await (await api.at(head.hash)).query.system.events();
      counter++;
      testLog
        .getLog()
        .info(
          `â†’ find on ${api.runtimeChain} for '${method}' event, attempt ${counter}, head ${head.hash}`,
        );

      events.forEach((e) => logEvent(api.runtimeChain, e));

      const filtered = _.filter(
        events,
        ({ event }) =>
          `${event.section}.${event.method}` === method &&
          (withData.length > 0
            ? JSON.stringify(event.data.toHuman()).includes(withData)
            : true),
      );
      if (filtered.length > 0) {
        resolve(filtered);
        unsub();
      }
      if (counter === blocks) {
        reject(`method ${method} not found within blocks limit`);
      }
    });
  });
};

export const waitForRewards = async (
  user: User,
  liquidityAssetId: BN,
  max: number = 20,
) =>
  new Promise(async (resolve) => {
    let numblocks = max;
    const unsub = await api.rpc.chain.subscribeNewHeads(async (header) => {
      numblocks--;
      const { chainUri } = getEnvironmentRequiredVars();
      const mangata = await getMangataInstance(chainUri);
      const price = await mangata.rpc.calculateRewardsAmount({
        address: user.keyRingPair.address,
        liquidityTokenId: liquidityAssetId.toString(),
      });

      if (price.gtn(0)) {
        unsub();
        resolve({});
      } else {
        testLog
          .getLog()
          .info(
            `#${header.number}  ${user.keyRingPair.address} (LP${liquidityAssetId}) - no rewards yet`,
          );
      }
      if (numblocks < 0) {
        unsub();
        reject(
          `Waited too long for rewards :( #${header.number}  ${user.keyRingPair.address} (LP${liquidityAssetId} `,
        );
      }
    });
  });

type CodecOrArray = Codec | Codec[];

const processCodecOrArray = (codec: CodecOrArray, fn: (c: Codec) => any) =>
  Array.isArray(codec) ? codec.map(fn) : fn(codec);

const toHuman = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toHuman?.() ?? c);
const toJson = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toJSON?.() ?? c);
const toHex = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toHex?.() ?? c);

export const matchSnapshot = (
  codec: CodecOrArray | Promise<CodecOrArray>,
  message?: string,
) => {
  return expect(Promise.resolve(codec).then(toHuman)).resolves.toMatchSnapshot(
    message,
  );
};

export const expectEvent = (codec: CodecOrArray, event: any) => {
  return expect(toHuman(codec)).toEqual(
    expect.arrayContaining([expect.objectContaining(event)]),
  );
};

export const expectHuman = (codec: CodecOrArray) => {
  return expect(toHuman(codec));
};

export const expectJson = (codec: CodecOrArray) => {
  return expect(toJson(codec));
};

export const expectHex = (codec: CodecOrArray) => {
  return expect(toHex(codec));
};

type EventFilter = string | { method: string; section: string };

const _matchEvents = async (
  msg: string,
  events: Promise<Codec[] | Codec>,
  ...filters: EventFilter[]
) => {
  let data = toHuman(await events).map(
    ({ event: { index: _, ...event } }: any) => event,
  );
  if (filters) {
    const filtersArr = Array.isArray(filters) ? filters : [filters];
    data = data.filter((evt: any) => {
      return filtersArr.some((filter) => {
        if (typeof filter === "string") {
          return evt.section === filter;
        }
        const { section, method } = filter;
        return evt.section === section && evt.method === method;
      });
    });
  }
  return expect(data).toMatchSnapshot(msg);
};

export const matchEvents = async (
  events: Promise<Codec[] | Codec> | Codec[],
  ...filters: EventFilter[]
) => {
  return _matchEvents("events", redact(events), ...filters);
};

export const matchSystemEvents = async (
  { api }: { api: ApiPromise },
  ...filters: EventFilter[]
) => {
  await _matchEvents(
    "system events",
    redact(api.query.system.events()),
    ...filters,
  );
};

export const matchUmp = async ({ api }: { api: ApiPromise }) => {
  expect(await api.query.parachainSystem.upwardMessages()).toMatchSnapshot(
    "ump",
  );
};

export const redact = async (data: any | Promise<any>) => {
  const json = toHuman(await data);

  const process = (obj: any): any => {
    if (obj == null) {
      return obj;
    }
    if (Array.isArray(obj)) {
      return obj.map(process);
    }
    if (typeof obj === "number") {
      return "(redacted)";
    }
    if (typeof obj === "string") {
      if (obj.match(/^[\d,]+$/) || obj.match(/0x[0-9a-f]{64}/)) {
        return "(redacted)";
      }
      return obj;
    }
    if (typeof obj === "object") {
      return Object.fromEntries(
        Object.entries(obj).map(([k, v]) => [k, process(v)]),
      );
    }
    return obj;
  };

  return process(json);
};

export const expectExtrinsicSuccess = (events: Codec[]) => {
  expectEvent(events, {
    event: expect.objectContaining({
      method: "ExtrinsicSuccess",
      section: "system",
    }),
  });
};
export const expectMGAExtrinsicSuDidSuccess = (
  events: MangataGenericEvent[],
) => {
  const anyError = events
    .filter((x) => x.method === "SudoAsDone" && x.section === "sudo")
    .some((x) => JSON.parse(JSON.stringify(x.eventData[0].data)).err);
  if (!anyError) {
    testLog.getLog().warn(JSON.stringify(events));
  }
  expect(anyError).toBeFalsy();
  return events!;
};
export const expectMGAExtrinsicSuDidFailed = (
  events: MangataGenericEvent[],
) => {
  const sudoErrorEvent = events.find(
    (x) =>
      x.method === "SudoAsDone" &&
      x.section === "sudo" &&
      JSON.parse(JSON.stringify(x.eventData[0].data)).err !== undefined,
  );
  expect(sudoErrorEvent).not.toBeNull();
  return sudoErrorEvent!;
};
