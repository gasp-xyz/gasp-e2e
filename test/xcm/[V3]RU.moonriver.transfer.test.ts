import { connectParachains } from "@acala-network/chopsticks";
import { AssetId } from "../../utils/ChainSpecs";
import { ApiContext } from "../../utils/Framework/XcmHelper";
import XcmNetworks from "../../utils/Framework/XcmNetworks";
import { alice, setupApi, setupUsers } from "../../utils/setup";
import {
  expectEvent,
  expectExtrinsicSuccess,
  expectJson,
  matchSystemEvents,
} from "../../utils/validators";
import { BN_BILLION, Mangata, BN_TEN, BN_HUNDRED } from "@mangata-finance/sdk";
import { mangataChopstick } from "../../utils/api";
import { Codec } from "@polkadot/types/types";
import { BN, BN_THOUSAND } from "@polkadot/util";
import { Keyring } from "@polkadot/api";
import { waitForEvents } from "../../utils/eventListeners";
import { testLog } from "../../utils/Logger";
import { jest } from "@jest/globals";

/**
 * @group skip-xcm
 * @group proxied
 */

jest.setTimeout(140000);

const keyring = new Keyring({ type: "ethereum" });
const alith = keyring.addFromUri(
  "0x5fb92d6e98884f76de468fa3f6278f8807c48bebc13595d45af5bdc4da702133",
  undefined,
  "ethereum",
);

describe("[V3][V3] XCM tests for Mangata <-> moonriver", () => {
  let moonriver: ApiContext;
  let mangata: ApiContext;

  beforeAll(async () => {
    await setupApi();
    moonriver = await XcmNetworks.moonriver();
    mangata = mangataChopstick!;
    await connectParachains([moonriver.chain, mangata.chain]);
    setupUsers();
    await mangata.dev.setStorage({
      Sudo: {
        Key: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      },
      Tokens: {
        Accounts: [
          [
            [alice.keyRingPair.address, { token: 39 }],
            { free: AssetId.Mgx.unit.mul(BN_BILLION).toString() },
          ],
          [
            [alice.keyRingPair.address, { token: 0 }],
            { free: AssetId.Mgx.unit.mul(BN_BILLION).toString() },
          ],
        ],
      },
    });
    await moonriver.dev.setStorage({
      System: {
        Account: [
          [
            ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac"],
            {
              data: { free: "100000000000000000000000" },
              providers: 1,
            },
          ],
        ],
      },
    });
    // await upgradeMangata(mangata);
  });

  beforeEach(async () => {
    await mangata.dev.setStorage({
      Sudo: {
        Key: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      },
      Tokens: {
        Accounts: [
          [
            [alice.keyRingPair.address, { token: 39 }],
            { free: AssetId.Mgx.unit.mul(BN_BILLION).toString() },
          ],
          [
            [alice.keyRingPair.address, { token: 0 }],
            { free: AssetId.Mgx.unit.mul(BN_BILLION).toString() },
          ],
        ],
      },
    });
    await moonriver.dev.setStorage({
      System: {
        Account: [
          [
            ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac"],
            { data: { free: "100000000000000000000000" }, providers: 1 },
          ],
        ],
      },
      TechCommitteeCollective: {
        Members: ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac"],
      },
      CouncilCollective: {
        Members: ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac"],
      },
      TreasuryCouncilCollective: {
        Members: ["0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac"],
      },
      AuthorFilter: {
        EligibleRatio: 100,
        EligibleCount: 100,
      },
    });
  });

  it("[V3] mangata transfer MGX assets to [V3] moonriver", async () => {
    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 0),
    ).toMatchSnapshot("Before");

    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.withdrawToMoonriver({
      account: alice.keyRingPair,
      amount: BN_HUNDRED.mul(BN_TEN.pow(new BN(18))),
      moonriverAddress: alith.address,
      tokenSymbol: "MGX",
      txOptions: {
        async extrinsicStatus(events) {
          expectExtrinsicSuccess(events as any[] as Codec[]);
          expectEvent(events as any as Codec[], {
            event: expect.objectContaining({
              section: "xTokens",
              method: "TransferredMultiAssets",
            }),
          });
        },
      },
    });

    await mangata.chain.newBlock();

    await moonriver.chain.newBlock();

    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 0),
    ).toMatchSnapshot("After");
    await matchSystemEvents(moonriver, "xcmpQueue", "Success");
  });

  it("[V3] moonriver transfer MGX assets to [V3] mangata", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    await mgaSdk.xTokens.withdrawToMoonriver({
      account: alice.keyRingPair,
      amount: BN_THOUSAND.mul(BN_TEN.pow(new BN(18))),
      moonriverAddress: alith.address,
      tokenSymbol: "MGX",
      txOptions: {
        async extrinsicStatus(events) {
          expectExtrinsicSuccess(events as any[] as Codec[]);
          expectEvent(events as any as Codec[], {
            event: expect.objectContaining({
              section: "xTokens",
              method: "TransferredMultiAssets",
            }),
          });
        },
      },
    });

    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 0),
    ).toMatchSnapshot("Before");

    const asset = {
      V3: {
        id: {
          Concrete: {
            parents: 1,
            interior: {
              X2: [
                {
                  Parachain: 2110,
                },
                {
                  GeneralKey: {
                    length: 4,
                    data: "0x0000000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              ],
            },
          },
        },
        fun: {
          Fungible: BN_HUNDRED.mul(BN_TEN.pow(new BN(18))),
        },
      },
    };

    const destination = {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: 2110,
            },
            {
              AccountId32: {
                id: moonriver.api
                  .createType(
                    "AccountId32",
                    "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY", // Alice address on Mangata
                  )
                  .toHex(),
              },
            },
          ],
        },
      },
    };
    const depositMgxArgs = {
      account: alith,
      url: moonriver.uri,
      asset,
      destination,
      weightLimit: "Unlimited",
    };
    await mgaSdk.xTokens.depositFromParachain(depositMgxArgs);

    //    await moonriver.api.tx.xTokens
    //      .transferMultiasset(asset, destination, "Unlimited")
    //      .signAndSend(alith);

    await moonriver.chain.newBlock();
    await mangata.chain.newBlock();

    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 0),
    ).toMatchSnapshot("After");
    await matchSystemEvents(mangata, "xcmpQueue", "Success");
  });

  it("[V3] mangata transfer MOVR assets to [V3] moonriver", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    //setup_add tokens from moonriver to mga.
    const asset = {
      V3: {
        id: {
          Concrete: {
            parents: 0,
            interior: {
              X1: {
                PalletInstance: 10,
              },
            },
          },
        },
        fun: {
          Fungible: BN_HUNDRED.mul(BN_TEN.pow(new BN(18))),
        },
      },
    };

    const destination = {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: 2110,
            },
            {
              AccountId32: {
                id: moonriver.api
                  .createType(
                    "AccountId32",
                    "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY", // Alice address on Mangata
                  )
                  .toHex(),
              },
            },
          ],
        },
      },
    };

    const depositMovrArgs = {
      account: alith,
      url: moonriver.uri,
      asset,
      destination,
      weightLimit: "Unlimited",
    };
    await mgaSdk.xTokens.depositFromParachain(depositMovrArgs);

    //    await moonriver.api.tx.xTokens
    //      .transferMultiasset(asset, destination, "Unlimited")
    //      .signAndSend(alith);
    await moonriver.chain.newBlock();
    await mangata.chain.newBlock();

    //act
    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 39),
    ).toMatchSnapshot("Before");
    await moonriver.chain.newBlock();
    await mgaSdk.xTokens.withdrawToMoonriver({
      account: alice.keyRingPair,
      amount: BN_TEN.mul(BN_TEN.pow(new BN(18))),
      moonriverAddress: alith.address,
      tokenSymbol: "MOVR",
      txOptions: {
        async extrinsicStatus(events) {
          expectExtrinsicSuccess(events as any[] as Codec[]);
          expectEvent(events as any as Codec[], {
            event: expect.objectContaining({
              section: "xTokens",
              method: "TransferredMultiAssets",
            }),
          });
        },
      },
    });

    await mangata.chain.newBlock();
    await moonriver.chain.newBlock();

    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 39),
    ).toMatchSnapshot("After");

    await matchSystemEvents(moonriver, "xcmpQueue", "Success");
  });

  it("[V3] moonriver transfer MOVR assets to [V3] mangata", async () => {
    const mgaSdk = Mangata.instance([mangata.uri]);
    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 39),
    ).toMatchSnapshot("Before");

    const asset = {
      V3: {
        id: {
          Concrete: {
            parents: 0,
            interior: {
              X1: {
                PalletInstance: 10,
              },
            },
          },
        },
        fun: {
          Fungible: BN_HUNDRED.mul(BN_TEN.pow(new BN(18))),
        },
      },
    };

    const destination = {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: 2110,
            },
            {
              AccountId32: {
                id: moonriver.api
                  .createType(
                    "AccountId32",
                    "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY", // Alice address on Mangata
                  )
                  .toHex(),
              },
            },
          ],
        },
      },
    };
    const depositMovrArgs = {
      account: alith,
      url: moonriver.uri,
      asset,
      destination,
      weightLimit: "Unlimited",
    };
    await mgaSdk.xTokens.depositFromParachain(depositMovrArgs);

    //    await moonriver.api.tx.xTokens
    //      .transferMultiasset(asset, destination, "Unlimited")
    //      .signAndSend(alith);
    await moonriver.chain.newBlock();
    testLog.getLog().info("Waiting for event2 - setup");
    await waitForEvents(mangata.api, "xcmpQueue.Success");
    testLog.getLog().info("done waiting ");
    expectEvent(await waitForEvents(mangata.api, "tokens.Deposited"), {
      event: expect.objectContaining({
        data: expect.objectContaining({
          who: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          amount: "99,998,387,200,000,000,000",
        }),
      }),
    });
    testLog.getLog().info("end expect");
    await mangata.chain.newBlock();
    expectJson(
      await mangata.api.query.tokens.accounts(alice.keyRingPair.address, 39),
    ).toMatchSnapshot("After");
  });
});
