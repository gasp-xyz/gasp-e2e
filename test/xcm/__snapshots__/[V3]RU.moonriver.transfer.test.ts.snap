// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] mangata transfer MGX assets to [V3] moonriver: After 1`] = `
{
  "free": "999999879880303252117536339",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] mangata transfer MGX assets to [V3] moonriver: Before 1`] = `
{
  "free": "0x00000000033b2e3c9fd0803ce8000000",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] mangata transfer MGX assets to [V3] moonriver: system events 1`] = `
[
  {
    "data": {
      "messageHash": "(redacted)",
      "messageId": "(redacted)",
      "weight": {
        "proofSize": "(redacted)",
        "refTime": "(redacted)",
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] mangata transfer MOVR assets to [V3] moonriver: After 1`] = `
{
  "free": "0x00000000033b2e4180cb00907e100000",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] mangata transfer MOVR assets to [V3] moonriver: Before 1`] = `
{
  "free": "0x00000000033b2e420b92239507f80000",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] mangata transfer MOVR assets to [V3] moonriver: system events 1`] = `
[
  {
    "data": {
      "messageHash": "(redacted)",
      "messageId": "(redacted)",
      "weight": {
        "proofSize": "(redacted)",
        "refTime": "(redacted)",
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] moonriver transfer MGX assets to [V3] mangata: After 1`] = `
{
  "free": "0x00000000033b2e0b55c187d9416b382f",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] moonriver transfer MGX assets to [V3] mangata: Before 1`] = `
{
  "free": "0x00000000033b2e0632892bc52602382f",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] moonriver transfer MGX assets to [V3] mangata: system events 1`] = `[]`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] moonriver transfer MOVR assets to [V3] mangata: After 1`] = `
{
  "free": "0x00000000033b2e420b92239507f80000",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`[V3][V3] XCM tests for Mangata <-> moonriver [V3] moonriver transfer MOVR assets to [V3] mangata: Before 1`] = `
{
  "free": "0x00000000033b2e3c9fd0803ce8000000",
  "frozen": 0,
  "reserved": 0,
}
`;
